src/python_testing/TC_IDM_4_2.py

```diff
@@ -0,0 +1,587 @@
+                autoResubscribe=False
+            )
+        except ChipStackError as e:
+            # Verify that the DUT returns an "INVALID_ACTION" status response
+            asserts.assert_equal(e.err, INVALID_ACTION_ERROR_CODE,
+                                 "Incorrect error response for subscription to unallowed cluster")
+
+        '''
+        ##########
+        Step 5
+        ##########
+        '''
+        self.print_step(5, "Setup CR2 such that it does not have access to an Endpoint. CR2 sends a subscription request to subscribe to all attributes on all clusters on a specific Endpoint for which it does not have access.")
+
+        # Limited ACE for controller 2 with endpoint 1 access only
+        CR2_limited_ace = Clusters.AccessControl.Structs.AccessControlEntryStruct(
+            privilege=Clusters.AccessControl.Enums.AccessControlEntryPrivilegeEnum.kView,
+            authMode=Clusters.AccessControl.Enums.AccessControlEntryAuthModeEnum.kCase,
+            targets=[Clusters.AccessControl.Structs.AccessControlTargetStruct(
+                endpoint=1,
+                cluster=Clusters.BasicInformation.id
+            )],
+            subjects=[CR2_nodeid])
+
+        # Add controller 2 limited ACE to DUT ACL
+        dut_acl = copy.deepcopy(dut_acl_original)
+        dut_acl.append(CR2_limited_ace)
+
+        # Write updated DUT ACL into DUT
+        await CR1.WriteAttribute(
+            nodeid=self.dut_node_id,
+            attributes=[(0, Clusters.AccessControl.Attributes.Acl(dut_acl))]
+        )
+
+        # Controller 2 tries to subscribe to all attributes from all clusters
+        # on an endpoint it doesn't have access to
+        # "INVALID_ACTION" status response expected
+        try:
+            await CR2.ReadAttribute(
+                nodeid=self.dut_node_id,
+                # Endpoint controller 2 has no access to
+                attributes=[(0)],
+                keepSubscriptions=False,
+                reportInterval=[3, 3],
+                autoResubscribe=False
+            )
+        except ChipStackError as e:
+            # Verify that the DUT returns an "INVALID_ACTION" status response
+            asserts.assert_equal(e.err, INVALID_ACTION_ERROR_CODE,
+                                 "Incorrect error response for subscription to unallowed endpoint")
+
+        '''
+        ##########
+        Step 6
+        ##########
+        '''
+        self.print_step(6, "Setup CR2 such that it does not have access to the Node. CR2 sends a subscription request to subscribe to all attributes on all clusters on all endpoints on a Node for which it does not have access.")
+
+        # ???????????????????????????????????????????
+        # Skip setting an ACE for controller 2 so the DUT node rejects subscribing to it?
+        # ???????????????????????????????????????????
+
+        # Add controller 2 limited ACE to DUT ACL
+        dut_acl = copy.deepcopy(dut_acl_original)
```

Comment:
You can just send the original directly, no need to copy.

Reply:
same here - I think this got missed.

Reply:
Updated